pandas/core/reshape/reshape.py

```diff
@@ -1089,6 +1089,198 @@ def get_empty_frame(data) -> DataFrame:
         return DataFrame(dummy_mat, index=index, columns=dummy_cols)
 
 
+def from_dummies(
+    data: DataFrame,
+    subset: None | list[str] = None,
+    sep: None | str | list[str] | dict[str, str] = None,
+    dropped_first: None | str | list[str] | dict[str, str] = None,
+) -> DataFrame:
+    """
+    Create a categorical `DataFrame` from a `DataFrame` of dummy variables.
+
+    Inverts the operation performed by 'get_dummies'.
+
+    Parameters
+    ----------
+    data : `DataFrame`
+        Data which contains dummy-coded variables.
+    subset : None, Index, or list of str, default 'None'
+        The columns which to convert from dummy-encoding and return as categorical
+        `DataFrame`.
+        If `columns` is None then all dummy columns are converted and appended
+        to the non-dummy columns.
+    sep : str, list of str, or dict of str, default '_'
+        Separator used in the column names of the dummy categories they are
+        character indicating the separation of the categorical names from the prefixes.
+        For example, if your column names are 'prefix_A' and 'prefix_B',
+        you can strip the underscore by specifying sep='_'.
+        Pass a list if multiple prefix separators are used in the columns names.
+        Will separate the prefix based on the first encountered separator following
+        the order of the list. Alternatively, pass a dictionary to map prefix
+        separators to prefixes if multiple and/or mixed separators are used in the
+        column names.
+    dropped_fist : None, str, list of str, or dict of str, default None
```

Comment:
Can you give an example of list or dict usage?  I can't why you need more than one.

Reply:
I think `str` is too narrow here as well.  BTW, what is dict of str?  Is the dict[Hashable, str], or are you suggesting dict[str, Number]?

Reply:
[An example where a `dict` is required can be found here.](https://github.com/pandas-dev/pandas/pull/41902#discussion_r705754802)
Technically only using `dict` could cover all possible scenarios, I just thought that in a case where the prefix separation is simple (but mixed) and when there are a lot of prefixes its less effort to just pass a list with the e.g. two required `sep` compared to passing a `dict` covering all prefixes.
```python
>>> dummies = DataFrame(
      {
          "col1_a": [1, 0, 1],
          "col1_b": [0, 1, 0],
          "col2-a": [0, 1, 0],
          "col2-b": [1, 0, 1],
          "col3_a": [0, 1, 0],
          "col3_b": [1, 0, 1],
          "col4-a": [0, 1, 0],
          "col4-b": [1, 0, 1],
          "col5_a": [0, 1, 0],
          "col5_b": [1, 0, 1],
          "col6-a": [0, 1, 0],
          "col6-b": [1, 0, 1],
          ...
      },
  )
  ```

Reply:
> technically only using dict could cover all possible scenarios

To be honest I'd stick to that - either a user passes a single separator (e.g. `'_'`) or a dict, no need for the list